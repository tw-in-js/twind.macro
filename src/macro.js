// @ts-check
const { createMacro } = require("babel-plugin-macros")
const { localName } = require("./constants")
const { raise } = require("./helpers")

module.exports = createMacro(function twindMacro({
	state,
	babel: { types: t },
}) {
	const program = state.file.path

	program.traverse({
		JSXOpeningElement(path) {
			const attributeList = path.get("attributes")

			/**
			 * @param {string} name
			 * @returns {babel.NodePath<babel.types.JSXAttribute> | undefined}
			 */
			const getAttribute = (name) => {
				for (const attribute of attributeList) {
					if (attribute.isJSXAttribute()) {
						const nodeName = attribute.node.name
						if (t.isJSXIdentifier(nodeName) && nodeName.name === name) {
							return attribute
						}
					}
				}
			}

			/**
			 * @param {babel.NodePath<babel.types.JSXAttribute>} attributePath
			 */
			const getAttributeValue = (attributePath) => {
				const attributeValue =
					attributePath.node.value ??
					raise(
						`Unexpected error: ` +
							`JSX attribute "${attributePath.node.name.name}" has no value`,
					)

				return t.isJSXExpressionContainer(attributeValue)
					? assertNotJSXEmpty(attributeValue.expression, t)
					: attributeValue
			}

			const twAttribute = getAttribute("tw")
			if (!twAttribute) return

			const twAttributeValue = getAttributeValue(twAttribute)
			const twCall = t.callExpression(t.identifier(localName), [
				twAttributeValue,
			])

			const classAttribute = getAttribute("className")

			const classAttributeValue =
				classAttribute && getAttributeValue(classAttribute)

			const newAttributeValue = classAttributeValue
				? t.templateLiteral(
						[
							t.templateElement({ raw: "" }),
							t.templateElement({ raw: " " }),
							t.templateElement({ raw: "" }),
						],
						[classAttributeValue, twCall],
				  )
				: twCall

			twAttribute.remove()
			classAttribute?.remove()

			path.node.attributes.push(
				t.jsxAttribute(
					t.jsxIdentifier("className"),
					t.jsxExpressionContainer(newAttributeValue),
				),
			)
		},
	})

	program.node.body.unshift(
		t.importDeclaration(
			[t.importSpecifier(t.identifier(localName), t.identifier("tw"))],
			t.stringLiteral("twind"),
		),
	)
})

/**
 * @template {babel.types.Node} T
 * @param {T} expression
 * @param {typeof babel.types} t
 * @returns {Exclude<T, babel.types.JSXEmptyExpression>}
 */
function assertNotJSXEmpty(expression, t) {
	if (t.isJSXEmptyExpression(expression)) {
		throw new Error("JSX empty expressions are not allowed")
	}
	// @ts-ignore
	return expression
}
