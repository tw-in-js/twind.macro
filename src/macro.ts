import * as babel from '@babel/core'
import annotateAsPure from '@babel/helper-annotate-as-pure'
import { createMacro, MacroParams } from 'babel-plugin-macros'
import { findJsxAttributeByName, getJsxAttributeName, getJsxAttributeValue } from './jsx-attribute'

function twindMacro({ state, references }: MacroParams) {
  const program = state.file.path
  const localImportName = program.scope.generateUid('tw')

  program.traverse({
    JSXOpeningElement(path) {
      const twAttribute = findJsxAttributeByName(path.node, 'tw')
      const twAttributeValue = getJsxAttributeValue(twAttribute)
      if (!twAttributeValue) return

      const twCall = babel.types.callExpression(babel.types.identifier(localImportName), [
        twAttributeValue,
      ])
      annotateAsPure(twCall)

      const classAttribute = findJsxAttributeByName(path.node, 'className')
      const classAttributeValue = getJsxAttributeValue(classAttribute)

      const newAttributeValue = classAttributeValue
        ? babel.types.templateLiteral(
            [
              babel.types.templateElement({ raw: '' }),
              babel.types.templateElement({ raw: ' ' }),
              babel.types.templateElement({ raw: '' }),
            ],
            [classAttributeValue, twCall],
          )
        : twCall

      path.node.attributes = path.node.attributes
        // remove existing tw and className attributes
        .filter((node) => {
          const name = getJsxAttributeName(node)
          return name !== 'tw' && name !== 'className'
        })
        // add new className attribute
        .concat(
          babel.types.jsxAttribute(
            babel.types.jsxIdentifier('className'),
            babel.types.jsxExpressionContainer(newAttributeValue),
          ),
        )
    },
  })

  references['tw']?.forEach((path) => {
    if (babel.types.isIdentifier(path.node)) {
      path.node.name = localImportName
    }
    annotateAsPure(path.parent)
  })

  program.node.body.unshift(
    babel.types.importDeclaration(
      [
        babel.types.importSpecifier(
          babel.types.identifier(localImportName),
          babel.types.identifier('tw'),
        ),
      ],
      babel.types.stringLiteral('twind'),
    ),
  )
}

export default createMacro(twindMacro)
