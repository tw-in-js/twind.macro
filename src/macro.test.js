// @ts-check
const { transform } = require("@babel/core")
const { join } = require("path")

test("tw prop - simple", () => {
	const input = `
	import "./macro"
	const simple = <button tw="bg-blue-500" />
	`

	expect(getAst(input)).toMatchSnapshot()
})

test("tw prop - complex", () => {
	const input = `
	import "./macro"
	const complex = (
		<button
			tw={[
				"bg-blue-500",
				condition && "text-white",
				{ 'leading-none': true },
			]}
		/>
	)
	`

	expect(getAst(input)).toMatchSnapshot()
})

/**
 * @param {string} code
 */
function getAst(code) {
	/** @type {import('@babel/core').TransformOptions} */
	const options = {
		presets: [],
		plugins: ["@babel/syntax-jsx", "macros"],
		filename: join(__dirname, "testfile.js"),
	}
	return transform(code, options).code
}
