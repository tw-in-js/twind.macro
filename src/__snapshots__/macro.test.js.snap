// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tw prop - complex 1`] = `
"import { tw } from \\"twind\\";
const complex = <button className={tw([\\"bg-blue-500\\", condition && \\"text-white\\", {
  'leading-none': true
}])} />;"
`;

exports[`tw prop - simple 1`] = `
"import { tw } from \\"twind\\";
const simple = <button className={tw(\\"bg-blue-500\\")} />;"
`;
